### PR TITLE
Safe Site Deletion

### PR DESCRIPTION
**Pull request by stephenrlouie**
### Change log

~- Will reject site deletions if services reference it.~

Updated: 7.5.17 1PM

- Mark for deletion flag added to Sites
- Garbage collection will occur every so often(1m), looking at the sites marked for deletion and ensure there are no references on them. Printing when references are found and when a deletion is successful
- Rejects Sites and Services referencing a site that is marked for deletion
- Exposed the deletion flag through the api

### TODO

- [x] block deleting the default site
- [x] this needs tests. 
- [x] ~default running site if we cannot find the desired site.~ Not an issue because we cannot delete sites if services reference them.
- [x] MARK FOR DELETION : Sites will be marked for deletion
   - [x] needs to block these sites from getting new site children
   - [x] needs to block services from using this site
   - [x] exposes the field in swagger
   - [x] ensure DB calls will properly handle Remove
   - [x] ensure converts will handle Remove
   - [x] Test queries on marked for delete sites

~@mikeynap some of the convert values changed on User Token. Can you look at the changes I made due to the re-generation of the client via swagger.~

### CODE REVIEW

File: api/v1/server/server.go [hunk 1]
```diff
@@ -1220,6 +1222,17 @@ func (s *Server) UpdateService(w http.ResponseWriter, r *http.Request) {
 		updatedService.MetaImpl = service2Update.MetaImpl
 		updatedService.Status = service2Update.Status
 		updatedService.Id = service2Update.Id
+
+		doc, err := s.db.GetSite(updatedService.Site)
+		if err != nil {
+			return http.StatusNotFound, fmt.Errorf("Could not get the services Site: %v", err)
```
Comment (vgunning) [inline api/v1/server/server.go]: Also yay.

File: api/v1/server/server_test.go [hunk 2]
```diff
@@ -181,6 +181,7 @@ func TestValidateService(t *testing.T) {
 	scrap.Name = "scrap"
 	scrap.Plugin = emptyPlugin.Name
 	scrap.PluginVersion = emptyPlugin.Version
+	scrap.Site = "local"
```
Comment (vgunning) [inline api/v1/server/server_test.go]: `model.DefaultSite`?

File: api/v1/server/server_test.go [hunk 3]
```diff
@@ -207,19 +208,18 @@ func TestValidateService(t *testing.T) {
...
 	}{
-		{scrap, fmt.Errorf(ErrDuplicateName)},
-		{noName, fmt.Errorf(ErrInvalidName)},
-		{badName, fmt.Errorf(ErrInvalidName)},
-		{unknownSite, fmt.Errorf(ErrUnknownSite)},
-		{noSite, fmt.Errorf(ErrUnknownSite)},
+		{scrap, ErrDuplicateName},
+		{noName, ErrInvalidName},
+		{badName, ErrInvalidName},
+		{unknownSite, nil},
```
Comment (vgunning) [inline api/v1/server/server_test.go]: These default to the `DefaultSite` now?

File: api/v1/server/server_test.go [hunk 4]
```diff
@@ -490,17 +494,106 @@ func TestDeleteService(t *testing.T) {
...
+		postSiteToAdd    apiclient.Site
+		siteErrCode      int
+		postServiceToAdd apiclient.Service
+		serviceErrCode   int
+	}{
+		{ // delete no parent
+			sitesToAdd:      []apiclient.Site{massSite, bxbSite, bosSite},
+			servicesToAdd:   []apiclient.Service{},
+			siteToDelete:    bosSite.Name,
+			expectedErrCode: 200,
```
Comment (vgunning) [inline api/v1/server/server_test.go]: Can we put in the `http` status codes from the library instead like `http.StatusOK`? maybe a bit more readable.

File: database/mock_database/mock_database.go [hunk 5]
```diff
@@ -1,14 +1,13 @@
...
 
 package mock_database
 
 import (
-	time "time"
-
 	gomock "github.com/golang/mock/gomock"
 	database "github.com/sigma-dev/sigma/database"
 	model "github.com/sigma-dev/sigma/model"
+	time "time"
```
Comment (vgunning) [inline database/mock_database/mock_database.go]: This is rather amusing.

File: database/nulldb.go [hunk 6]
```diff
@@ -422,7 +422,7 @@ func (n *NullDB) QuerySites(o model.Site) ([]model.Site, error) {
 			mdMatch = mdMatch && mdv == v.Metadata[k]
 		}
 
-		if mdMatch && (o.Id == "" || o.Id == v.Id) && (o.Parent == "" || o.Parent == v.Parent) && (o.Name == "" || o.Name == v.Name) {
+		if mdMatch && (o.Id == "" || o.Id == v.Id) && (o.Name == "" || o.Name == v.Name) && (o.Parent == "" || o.Parent == v.Parent) && (o.Remove == nil || model.BoolValue(o.Remove) == model.BoolValue(v.Remove)) {
```
Comment (vgunning) [inline database/nulldb.go]: Would it be better to make this huge if statement a tiny helper function?

File: manager/manager.go [hunk 7]
```diff
@@ -161,6 +172,10 @@ func (m *manager) Start() error {
 			m.workerWg.Wait()
 			log.Info("Quitting Sigma Manager")
 			return nil
+		case <-gcTicker.C:
+			// garbage collect Sites
+			log.Debug("Garbage collecting sites")
+			m.cleanSites()
 		case <-ticker.C:
```
Comment (vgunning) [inline manager/manager.go]: NIT: Rename ticker to its purpose? `svcTicker`?

File: manager/manager.go [hunk 8]
```diff
@@ -161,6 +172,10 @@ func (m *manager) Start() error {
 			m.workerWg.Wait()
 			log.Info("Quitting Sigma Manager")
 			return nil
+		case <-gcTicker.C:
+			// garbage collect Sites
+			log.Debug("Garbage collecting sites")
+			m.cleanSites()
```
Comment (vgunning) [inline manager/manager.go]: Do we want to log errors that are returned by the `m.cleanSites()`? or is it for testing purposes only

File: manager/manager.go [hunk 9]
```diff
@@ -205,6 +219,59 @@ func (m *manager) Stop() {
...
+		return err
+	}
+	for _, id := range siteIDList {
+		//check if I have children
+		myself, err := m.db.GetSite(id)
+		if err != nil {
+			return err
+		}
+
+		if myself.Name != model.DefaultSite {
```
Comment (vgunning) [inline manager/manager.go]: simplify logic/nesting by changing to 
```
if myself.Name == model.DefaultSite {
   continue
```

File: manager/manager.go [hunk 10]
```diff
@@ -205,6 +219,59 @@ func (m *manager) Stop() {
...
+// Looks for sites marked for deletion, deletes ones that are clear of children
+func (m *manager) cleanSites() error {
+	siteIDList, err := m.db.QuerySitesIds(model.Site{Remove: model.NewBool(true)})
+	if err != nil {
+		return err
+	}
+	for _, id := range siteIDList {
+		//check if I have children
+		myself, err := m.db.GetSite(id)
+		if err != nil {
```
Comment (vgunning) [inline manager/manager.go]: Do we want to completely error out if we fail to get one site by its `id`?

File: manager/manager_test.go [hunk 11]
```diff
@@ -256,6 +256,121 @@ func expectStartOperation(t *testing.T, m *manager, db *mock_database.MockDataba
 	tskr.EXPECT().Run(&op).Return(nil)
 }
 
+// a nil valie indicates that the corresponding call will not be executed by cleanSites
+func TestCleanSites(t *testing.T) {
```
Comment (vgunning) [inline manager/manager_test.go]: Have something with a mix of references by `id` and `name`?

File: api/v1/server/server_test.go (hunk 3 above)
Comment (stephenrlouie) [inline api/v1/server/server_test.go]: yes

File: api/v1/server/server_test.go (hunk 4 above)
Comment (stephenrlouie) [inline api/v1/server/server_test.go]: truth

File: manager/manager.go (hunk 8 above)
Comment (stephenrlouie) [inline manager/manager.go]: will log

File: manager/manager.go (hunk 10 above)
Comment (stephenrlouie) [inline manager/manager.go]: nah we will continue to try the others

File: manager/manager_test.go (hunk 11 above)
Comment (stephenrlouie) [inline manager/manager_test.go]: yup

File: database/nulldb.go (hunk 6 above)
Comment (stephenrlouie) [inline database/nulldb.go]: @mikeynap's PR addresses: #196

**Pull request merged**